{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BS1140 SCENIC+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action = 'ignore', category=FutureWarning)\n",
    "import sys\n",
    "import os\n",
    "import imp\n",
    "import dill\n",
    "import scanpy as sc\n",
    "import pandas\n",
    "import pyranges\n",
    "# set stderr to null to avoid strange messages from ray\n",
    "import sys\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull, 'wb')\n",
    "work_dir = 'BS1140'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00) \\n[Clang 13.0.1 ]'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sys.version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inferring enhancer-driven GRNs using SCENIC+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/73/t_19bmpd5m7bnvd_88sxp6lw0000gn/T/ipykernel_78415/1337946795.py:1: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses\n",
      "  import imp\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_h5ad(os.path.join(work_dir, 'scRNA/adata_archr.h5ad')) # preprocessed scRNA result\n",
    "cistopic_obj = dill.load(open(os.path.join(work_dir, 'scATAC/cistopic_obj.pkl'), 'rb')) # preprocessed scATAC results\n",
    "menr = dill.load(open(os.path.join(work_dir, 'motifs_euler/menr.pkl'), 'rb')) # preprocessed scATAC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-11-01 10:53:02,389 cisTopic     INFO     Imputing drop-outs\n",
      "2022-11-01 10:53:03,876 cisTopic     INFO     Scaling\n",
      "2022-11-01 10:53:04,858 cisTopic     INFO     Keep non zero rows\n",
      "2022-11-01 10:53:05,622 cisTopic     INFO     Imputed accessibility sparsity: 0.31028790911462056\n",
      "2022-11-01 10:53:05,622 cisTopic     INFO     Create CistopicImputedFeatures object\n",
      "2022-11-01 10:53:05,623 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 2546 x 19979 and n_cells x n_regions = 2546 x 154223\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'gene_ids', 'feature_types', 'genome', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "\tmetadata_cell:'GEX_n_genes', 'GEX_doublet_score', 'GEX_predicted_doublet', 'GEX_n_genes_by_counts', 'GEX_total_counts', 'GEX_total_counts_mt', 'GEX_pct_counts_mt', 'GEX_louvain_res_1.0', 'GEX_celltype', 'GEX_cluster_archr', 'ACC_Total_nr_frag', 'ACC_cisTopic_log_nr_frag', 'ACC_barcode', 'ACC_Total_nr_frag_in_regions', 'ACC_Log_total_nr_frag', 'ACC_Log_unique_nr_frag', 'ACC_cisTopic_nr_frag', 'ACC_Dupl_rate', 'ACC_Dupl_nr_frag', 'ACC_cisTopic_log_nr_acc', 'ACC_FRIP', 'ACC_TSS_enrichment', 'ACC_cisTopic_nr_acc', 'ACC_Unique_nr_frag', 'ACC_Unique_nr_frag_in_regions', 'ACC_n_genes', 'ACC_doublet_score', 'ACC_predicted_doublet', 'ACC_n_genes_by_counts', 'ACC_total_counts', 'ACC_total_counts_mt', 'ACC_pct_counts_mt', 'ACC_louvain_res_1.0', 'ACC_celltype', 'ACC_cluster_archr', 'ACC_sample_id'\n",
       "\tmenr:'CTX_topics_otsu_All', 'CTX_topics_otsu_No_promoters', 'DEM_topics_otsu_All', 'DEM_topics_otsu_No_promoters', 'CTX_topics_top_3_All', 'CTX_topics_top_3_No_promoters', 'DEM_topics_top_3_All', 'DEM_topics_top_3_No_promoters', 'CTX_DARs_All', 'CTX_DARs_No_promoters', 'DEM_DARs_All', 'DEM_DARs_No_promoters'\n",
       "\tdr_cell:'GEX_X_pca', 'GEX_X_umap'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create a SCENIC+ object to store gene expression, chromatin accessibility, motif enrichment results and cell/region/gene metadata\n",
    "from scenicplus.scenicplus_class import create_SCENICPLUS_object\n",
    "import numpy as np\n",
    "scplus_obj = create_SCENICPLUS_object(\n",
    "    GEX_anndata=adata.raw.to_adata(),\n",
    "    cisTopic_obj=cistopic_obj,\n",
    "    menr = menr,\n",
    "    bc_transform_func=lambda x: f'{x}-BS1140' # function to convert scATAC barcodes to scRNA ones\n",
    ")\n",
    "scplus_obj.X_EXP = np.array(scplus_obj.X_EXP.todense())\n",
    "scplus_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "host: 105\n",
      "Genes recovered: 15623 out of 19979\n",
      "host: 104\n",
      "Genes recovered: 15735 out of 19979\n",
      "host: 103\n",
      "Genes recovered: 19516 out of 19979\n",
      "host: 102\n",
      "Genes recovered: 19556 out of 19979\n",
      "host: 101\n",
      "Genes recovered: 19623 out of 19979\n",
      "host: 100\n",
      "Genes recovered: 19725 out of 19979\n",
      "host: 99\n",
      "Genes recovered: 19760 out of 19979\n",
      "host: 98\n",
      "Genes recovered: 19957 out of 19979\n",
      "host: 97\n",
      "Genes recovered: 19753 out of 19979\n",
      "host: 96\n",
      "Genes recovered: 19234 out of 19979\n",
      "host: 95\n",
      "Genes recovered: 19117 out of 19979\n",
      "host: 94\n",
      "Genes recovered: 19059 out of 19979\n",
      "host: 93\n",
      "Genes recovered: 18940 out of 19979\n",
      "host: 92\n",
      "Genes recovered: 18856 out of 19979\n",
      "host: 91\n",
      "Genes recovered: 18720 out of 19979\n",
      "host: 90\n",
      "Genes recovered: 18700 out of 19979\n",
      "host: 89\n",
      "Host not reachable\n",
      "host: 88\n",
      "Host not reachable\n",
      "host: 87\n",
      "Host not reachable\n",
      "host: 86\n",
      "Host not reachable\n",
      "host: 80\n",
      "Genes recovered: 15229 out of 19979\n",
      "host: 77\n",
      "Genes recovered: 15021 out of 19979\n",
      "host: 75\n",
      "Host not reachable\n",
      "host: 54\n",
      "Host not reachable\n",
      "version: 98 has the largest overlap, use http://sep2019.archive.ensembl.org/ as biomart host\n"
     ]
    }
   ],
   "source": [
    "# select a optimal biomart host\n",
    "ensembl_version_dict = {'105': 'http://www.ensembl.org',\n",
    "                        '104': 'http://may2021.archive.ensembl.org/',\n",
    "                        '103': 'http://feb2021.archive.ensembl.org/',\n",
    "                        '102': 'http://nov2020.archive.ensembl.org/',\n",
    "                        '101': 'http://aug2020.archive.ensembl.org/',\n",
    "                        '100': 'http://apr2020.archive.ensembl.org/',\n",
    "                        '99': 'http://jan2020.archive.ensembl.org/',\n",
    "                        '98': 'http://sep2019.archive.ensembl.org/',\n",
    "                        '97': 'http://jul2019.archive.ensembl.org/',\n",
    "                        '96': 'http://apr2019.archive.ensembl.org/',\n",
    "                        '95': 'http://jan2019.archive.ensembl.org/',\n",
    "                        '94': 'http://oct2018.archive.ensembl.org/',\n",
    "                        '93': 'http://jul2018.archive.ensembl.org/',\n",
    "                        '92': 'http://apr2018.archive.ensembl.org/',\n",
    "                        '91': 'http://dec2017.archive.ensembl.org/',\n",
    "                        '90': 'http://aug2017.archive.ensembl.org/',\n",
    "                        '89': 'http://may2017.archive.ensembl.org/',\n",
    "                        '88': 'http://mar2017.archive.ensembl.org/',\n",
    "                        '87': 'http://dec2016.archive.ensembl.org/',\n",
    "                        '86': 'http://oct2016.archive.ensembl.org/',\n",
    "                        '80': 'http://may2015.archive.ensembl.org/',\n",
    "                        '77': 'http://oct2014.archive.ensembl.org/',\n",
    "                        '75': 'http://feb2014.archive.ensembl.org/',\n",
    "                        '54': 'http://may2009.archive.ensembl.org/'}\n",
    "\n",
    "import pybiomart as pbm\n",
    "def test_ensembl_host(scplus_obj, host, species):\n",
    "    dataset = pbm.Dataset(name=species+'_gene_ensembl',  host=host)\n",
    "    annot = dataset.query(attributes=['chromosome_name', 'transcription_start_site', 'strand', 'external_gene_name', 'transcript_biotype'])\n",
    "    annot.columns = ['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "    annot['Chromosome'] = annot['Chromosome'].astype('str')\n",
    "    filter = annot['Chromosome'].str.contains('CHR|GL|JH|MT')\n",
    "    annot = annot[~filter]\n",
    "    annot.columns=['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "    gene_names_release = set(annot['Gene'].tolist())\n",
    "    ov=len([x for x in scplus_obj.gene_names if x in gene_names_release])\n",
    "    print('Genes recovered: ' + str(ov) + ' out of ' + str(len(scplus_obj.gene_names)))\n",
    "    return ov\n",
    "\n",
    "n_overlap = {}\n",
    "for version in ensembl_version_dict.keys():\n",
    "    print(f'host: {version}')\n",
    "    try:\n",
    "        n_overlap[version] =  test_ensembl_host(scplus_obj, ensembl_version_dict[version], 'hsapiens')\n",
    "    except:\n",
    "        print('Host not reachable')\n",
    "v = sorted(n_overlap.items(), key=lambda item: item[1], reverse=True)[0][0]\n",
    "print(f\"version: {v} has the largest overlap, use {ensembl_version_dict[v]} as biomart host\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the biomart host\n",
    "biomart_host = \"http://sep2019.archive.ensembl.org/\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "download list of known human TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "!wget -O pbmc_tutorial/data/utoronto_human_tfs_v_1.01.txt  http://humantfs.ccbr.utoronto.ca/download/v_1.01/TF_names_v_1.01.txt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " download a the program bedToBigBed this will be used to generate files which can be uploaded to the UCSC genome browser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "!wget -O pbmc_tutorial/bedToBigBed http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedToBigBed\n",
    "!chmod +x pbmc_tutorial/bedToBigBed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#only keep the first two columns of the PCA embedding in order to be able to visualize this in SCope\n",
    "scplus_obj.dr_cell['GEX_X_pca'] = scplus_obj.dr_cell['GEX_X_pca'].iloc[:, 0:2]\n",
    "#scplus_obj.dr_cell['GEX_rep'] = scplus_obj.dr_cell['GEX_rep'].iloc[:, 0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-11-01 13:23:37,372 SCENIC+_wrapper INFO     BS1140/scenicplus folder already exists.\n",
      "2022-11-01 13:23:37,374 SCENIC+_wrapper INFO     Inferring region to gene relationships\n",
      "2022-11-01 13:23:37,724 R2G          INFO     Calculating region to gene importances, using GBM method\n",
      "2022-11-01 13:39:04,143 R2G          INFO     Took 926.4185910224915 seconds\n",
      "2022-11-01 13:39:04,144 R2G          INFO     Calculating region to gene correlation, using SR method\n",
      "2022-11-01 13:46:53,149 R2G          INFO     Took 469.00419187545776 seconds\n",
      "2022-11-01 13:46:58,065 R2G          INFO     Done!\n",
      "2022-11-01 13:46:58,221 SCENIC+_wrapper INFO     Inferring TF to gene relationships\n",
      "2022-11-01 13:47:04,026 TF2G         INFO     Calculating TF to gene correlation, using GBM method\n"
     ]
    }
   ],
   "source": [
    "# computation done on cluster\n",
    "tmp_dir = '/Users/jinhuixin/tmp'\n",
    "from scenicplus.wrappers.run_scenicplus import run_scenicplus\n",
    "try:\n",
    "    run_scenicplus(\n",
    "        scplus_obj = scplus_obj,\n",
    "        variable = ['GEX_celltype'],\n",
    "        species = 'hsapiens',\n",
    "        assembly = 'hg38',\n",
    "        tf_file = '/Users/jinhuixin/Master/thesis/GRN/pbmc_tutorial/data/utoronto_human_tfs_v_1.01.txt',\n",
    "        save_path = os.path.join(work_dir, 'scenicplus'),\n",
    "        biomart_host = biomart_host,\n",
    "        upstream = [1000, 150000],\n",
    "        downstream = [1000, 150000],\n",
    "        calculate_TF_eGRN_correlation = True,\n",
    "        calculate_DEGs_DARs = True,\n",
    "        export_to_loom_file = True,\n",
    "        export_to_UCSC_file = True,\n",
    "        path_bedToBigBed = '/Users/jinhuixin/Master/thesis/GRN/pbmc_tutorial',\n",
    "        n_cpu = 7,\n",
    "        _temp_dir = os.path.join(tmp_dir, 'ray_spill'))\n",
    "except Exception as e:\n",
    "    #in case of failure, still save the object\n",
    "    dill.dump(scplus_obj, open(os.path.join(work_dir, 'scenicplus/scplus_obj.pkl'), 'wb'), protocol=-1)\n",
    "    raise(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.\n"
     ]
    }
   ],
   "source": [
    "# load scenicplus object generated on euler\n",
    "scplus_obj = dill.load(open(os.path.join(work_dir, 'scenicplus/scplus_obj.pkl'), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 2546 x 19979 and n_cells x n_regions = 2546 x 154223\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'gene_ids', 'feature_types', 'genome', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "\tmetadata_cell:'GEX_n_genes', 'GEX_doublet_score', 'GEX_predicted_doublet', 'GEX_n_genes_by_counts', 'GEX_total_counts', 'GEX_total_counts_mt', 'GEX_pct_counts_mt', 'GEX_louvain_res_1.0', 'GEX_celltype', 'GEX_cluster_archr', 'ACC_Total_nr_frag', 'ACC_cisTopic_log_nr_frag', 'ACC_barcode', 'ACC_Total_nr_frag_in_regions', 'ACC_Log_total_nr_frag', 'ACC_Log_unique_nr_frag', 'ACC_cisTopic_nr_frag', 'ACC_Dupl_rate', 'ACC_Dupl_nr_frag', 'ACC_cisTopic_log_nr_acc', 'ACC_FRIP', 'ACC_TSS_enrichment', 'ACC_cisTopic_nr_acc', 'ACC_Unique_nr_frag', 'ACC_Unique_nr_frag_in_regions', 'ACC_n_genes', 'ACC_doublet_score', 'ACC_predicted_doublet', 'ACC_n_genes_by_counts', 'ACC_total_counts', 'ACC_total_counts_mt', 'ACC_pct_counts_mt', 'ACC_louvain_res_1.0', 'ACC_celltype', 'ACC_cluster_archr', 'ACC_sample_id'\n",
       "\tmenr:'CTX_topics_otsu_All', 'CTX_topics_otsu_No_promoters', 'DEM_topics_otsu_All', 'DEM_topics_otsu_No_promoters', 'CTX_topics_top_3_All', 'CTX_topics_top_3_No_promoters', 'DEM_topics_top_3_All', 'DEM_topics_top_3_No_promoters', 'CTX_DARs_All', 'CTX_DARs_No_promoters', 'DEM_DARs_All', 'DEM_DARs_No_promoters'\n",
       "\tdr_cell:'GEX_X_pca', 'GEX_X_umap', 'eRegulons_UMAP', 'eRegulons_tSNE'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Output of SCENIC+"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene expression and chromatin accessibility data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.5</th>\n",
       "      <th>LINC01409</th>\n",
       "      <th>LINC01128</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>FAM41C</th>\n",
       "      <th>AL645608.6</th>\n",
       "      <th>SAMD11</th>\n",
       "      <th>NOC2L</th>\n",
       "      <th>KLHL17</th>\n",
       "      <th>...</th>\n",
       "      <th>MT-ND4</th>\n",
       "      <th>MT-ND5</th>\n",
       "      <th>MT-ND6</th>\n",
       "      <th>MT-CYB</th>\n",
       "      <th>MAFIP</th>\n",
       "      <th>AC011043.1</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AL592183.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "      <th>AC007325.4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATTCCTCCATAATCAC-1-BS1140</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.194697</td>\n",
       "      <td>1.194697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.194697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGAAACTGTTAGAGGG-1-BS1140</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.023663</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.648440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGACATCAGTTTGTG-1-BS1140</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.772823</td>\n",
       "      <td>2.140287</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.305294</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGCACTTGTTAGGTGC-1-BS1140</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.440718</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCAAACCAGGCAAGC-1-BS1140</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.628536</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19979 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           AL627309.1  AL627309.5  LINC01409  LINC01128  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140         0.0         0.0        0.0        0.0   \n",
       "TGAAACTGTTAGAGGG-1-BS1140         0.0         0.0        0.0        0.0   \n",
       "TTGACATCAGTTTGTG-1-BS1140         0.0         0.0        0.0        0.0   \n",
       "TGCACTTGTTAGGTGC-1-BS1140         0.0         0.0        0.0        0.0   \n",
       "CCCAAACCAGGCAAGC-1-BS1140         0.0         0.0        0.0        0.0   \n",
       "\n",
       "                           LINC00115  FAM41C  AL645608.6  SAMD11  NOC2L  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140        0.0     0.0         0.0     0.0    0.0   \n",
       "TGAAACTGTTAGAGGG-1-BS1140        0.0     0.0         0.0     0.0    0.0   \n",
       "TTGACATCAGTTTGTG-1-BS1140        0.0     0.0         0.0     0.0    0.0   \n",
       "TGCACTTGTTAGGTGC-1-BS1140        0.0     0.0         0.0     0.0    0.0   \n",
       "CCCAAACCAGGCAAGC-1-BS1140        0.0     0.0         0.0     0.0    0.0   \n",
       "\n",
       "                           KLHL17  ...    MT-ND4    MT-ND5  MT-ND6    MT-CYB  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140     0.0  ...  1.194697  1.194697     0.0  1.194697   \n",
       "TGAAACTGTTAGAGGG-1-BS1140     0.0  ...  0.000000  2.023663     0.0  2.648440   \n",
       "TTGACATCAGTTTGTG-1-BS1140     0.0  ...  2.772823  2.140287     0.0  3.305294   \n",
       "TGCACTTGTTAGGTGC-1-BS1140     0.0  ...  0.000000  0.000000     0.0  1.440718   \n",
       "CCCAAACCAGGCAAGC-1-BS1140     0.0  ...  0.000000  0.000000     0.0  2.628536   \n",
       "\n",
       "                           MAFIP  AC011043.1  AL354822.1  AL592183.1  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140    0.0         0.0         0.0         0.0   \n",
       "TGAAACTGTTAGAGGG-1-BS1140    0.0         0.0         0.0         0.0   \n",
       "TTGACATCAGTTTGTG-1-BS1140    0.0         0.0         0.0         0.0   \n",
       "TGCACTTGTTAGGTGC-1-BS1140    0.0         0.0         0.0         0.0   \n",
       "CCCAAACCAGGCAAGC-1-BS1140    0.0         0.0         0.0         0.0   \n",
       "\n",
       "                           AC240274.1  AC007325.4  \n",
       "ATTCCTCCATAATCAC-1-BS1140         0.0         0.0  \n",
       "TGAAACTGTTAGAGGG-1-BS1140         0.0         0.0  \n",
       "TTGACATCAGTTTGTG-1-BS1140         0.0         0.0  \n",
       "TGCACTTGTTAGGTGC-1-BS1140         0.0         0.0  \n",
       "CCCAAACCAGGCAAGC-1-BS1140         0.0         0.0  \n",
       "\n",
       "[5 rows x 19979 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.to_df('EXP').head() # expression counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ATTCCTCCATAATCAC-1-BS1140</th>\n",
       "      <th>TGAAACTGTTAGAGGG-1-BS1140</th>\n",
       "      <th>TTGACATCAGTTTGTG-1-BS1140</th>\n",
       "      <th>TGCACTTGTTAGGTGC-1-BS1140</th>\n",
       "      <th>CCCAAACCAGGCAAGC-1-BS1140</th>\n",
       "      <th>CACTGACCAAGTGTCC-1-BS1140</th>\n",
       "      <th>GATTTGCAGGTGTCCA-1-BS1140</th>\n",
       "      <th>TTAGCAGGTTTAGTCC-1-BS1140</th>\n",
       "      <th>AGGTAACCATCCCTCA-1-BS1140</th>\n",
       "      <th>GCGAAGCCAACTAGCC-1-BS1140</th>\n",
       "      <th>...</th>\n",
       "      <th>CGCCTCATCACTAAGC-1-BS1140</th>\n",
       "      <th>CAAATCATCTGCAACG-1-BS1140</th>\n",
       "      <th>CATAGCTAGGGTGAAC-1-BS1140</th>\n",
       "      <th>CTTCACTCAATGAGGT-1-BS1140</th>\n",
       "      <th>GTGCTGATCGCAAACT-1-BS1140</th>\n",
       "      <th>CGCATGATCCTCACTA-1-BS1140</th>\n",
       "      <th>GCTTTCATCATCAGTA-1-BS1140</th>\n",
       "      <th>AGCACTAGTACCGAAC-1-BS1140</th>\n",
       "      <th>GGGCAATAGGTAAGGC-1-BS1140</th>\n",
       "      <th>GCAAGTCGTGTTTCAC-1-BS1140</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GL000194.1:114742-115242</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000194.1:101154-101654</th>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000195.1:92476-92976</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000195.1:131013-131513</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000195.1:30584-31084</th>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>22</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2546 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ATTCCTCCATAATCAC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          7   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           16   \n",
       "\n",
       "                          TGAAACTGTTAGAGGG-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          3   \n",
       "GL000194.1:101154-101654                          9   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           13   \n",
       "\n",
       "                          TTGACATCAGTTTGTG-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          7   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           17   \n",
       "\n",
       "                          TGCACTTGTTAGGTGC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          8   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           15   \n",
       "\n",
       "                          CCCAAACCAGGCAAGC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          6   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           18   \n",
       "\n",
       "                          CACTGACCAAGTGTCC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          3   \n",
       "GL000194.1:101154-101654                         17   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           12   \n",
       "\n",
       "                          GATTTGCAGGTGTCCA-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          7   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          1   \n",
       "GL000195.1:30584-31084                           16   \n",
       "\n",
       "                          TTAGCAGGTTTAGTCC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          7   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           15   \n",
       "\n",
       "                          AGGTAACCATCCCTCA-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          5   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          1   \n",
       "GL000195.1:30584-31084                           21   \n",
       "\n",
       "                          GCGAAGCCAACTAGCC-1-BS1140  ...  \\\n",
       "GL000194.1:114742-115242                          2  ...   \n",
       "GL000194.1:101154-101654                          4  ...   \n",
       "GL000195.1:92476-92976                            0  ...   \n",
       "GL000195.1:131013-131513                          1  ...   \n",
       "GL000195.1:30584-31084                           21  ...   \n",
       "\n",
       "                          CGCCTCATCACTAAGC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          3   \n",
       "GL000194.1:101154-101654                         11   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           13   \n",
       "\n",
       "                          CAAATCATCTGCAACG-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          5   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           20   \n",
       "\n",
       "                          CATAGCTAGGGTGAAC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                         10   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           13   \n",
       "\n",
       "                          CTTCACTCAATGAGGT-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          7   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          1   \n",
       "GL000195.1:30584-31084                           20   \n",
       "\n",
       "                          GTGCTGATCGCAAACT-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          3   \n",
       "GL000194.1:101154-101654                          8   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           15   \n",
       "\n",
       "                          CGCATGATCCTCACTA-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          5   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          1   \n",
       "GL000195.1:30584-31084                           22   \n",
       "\n",
       "                          GCTTTCATCATCAGTA-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          7   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           17   \n",
       "\n",
       "                          AGCACTAGTACCGAAC-1-BS1140  \\\n",
       "GL000194.1:114742-115242                          2   \n",
       "GL000194.1:101154-101654                          6   \n",
       "GL000195.1:92476-92976                            0   \n",
       "GL000195.1:131013-131513                          0   \n",
       "GL000195.1:30584-31084                           17   \n",
       "\n",
       "                          GGGCAATAGGTAAGGC-1-BS1140  GCAAGTCGTGTTTCAC-1-BS1140  \n",
       "GL000194.1:114742-115242                          3                          2  \n",
       "GL000194.1:101154-101654                          9                          5  \n",
       "GL000195.1:92476-92976                            0                          0  \n",
       "GL000195.1:131013-131513                          0                          0  \n",
       "GL000195.1:30584-31084                           15                         20  \n",
       "\n",
       "[5 rows x 2546 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.to_df('ACC').head() # accessibility"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell, region and gene metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEX_n_genes</th>\n",
       "      <th>GEX_doublet_score</th>\n",
       "      <th>GEX_predicted_doublet</th>\n",
       "      <th>GEX_n_genes_by_counts</th>\n",
       "      <th>GEX_total_counts</th>\n",
       "      <th>GEX_total_counts_mt</th>\n",
       "      <th>GEX_pct_counts_mt</th>\n",
       "      <th>GEX_louvain_res_1.0</th>\n",
       "      <th>GEX_celltype</th>\n",
       "      <th>GEX_cluster_archr</th>\n",
       "      <th>...</th>\n",
       "      <th>ACC_doublet_score</th>\n",
       "      <th>ACC_predicted_doublet</th>\n",
       "      <th>ACC_n_genes_by_counts</th>\n",
       "      <th>ACC_total_counts</th>\n",
       "      <th>ACC_total_counts_mt</th>\n",
       "      <th>ACC_pct_counts_mt</th>\n",
       "      <th>ACC_louvain_res_1.0</th>\n",
       "      <th>ACC_celltype</th>\n",
       "      <th>ACC_cluster_archr</th>\n",
       "      <th>ACC_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATTCCTCCATAATCAC-1-BS1140</th>\n",
       "      <td>2069.0</td>\n",
       "      <td>0.031680</td>\n",
       "      <td>False</td>\n",
       "      <td>2068.0</td>\n",
       "      <td>4343.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.253281</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.031680</td>\n",
       "      <td>False</td>\n",
       "      <td>2068.0</td>\n",
       "      <td>4343.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.253281</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>BS1140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGAAACTGTTAGAGGG-1-BS1140</th>\n",
       "      <td>969.0</td>\n",
       "      <td>0.072093</td>\n",
       "      <td>False</td>\n",
       "      <td>967.0</td>\n",
       "      <td>1523.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.247538</td>\n",
       "      <td>Ex</td>\n",
       "      <td>Ex</td>\n",
       "      <td>C2-Exhausted-2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.072093</td>\n",
       "      <td>False</td>\n",
       "      <td>967.0</td>\n",
       "      <td>1523.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.247538</td>\n",
       "      <td>Ex</td>\n",
       "      <td>C2-Exhausted-2</td>\n",
       "      <td>C2-Exhausted-2</td>\n",
       "      <td>BS1140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGACATCAGTTTGTG-1-BS1140</th>\n",
       "      <td>1413.0</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>False</td>\n",
       "      <td>1409.0</td>\n",
       "      <td>2666.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>2.775694</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>False</td>\n",
       "      <td>1409.0</td>\n",
       "      <td>2666.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>2.775694</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>BS1140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGCACTTGTTAGGTGC-1-BS1140</th>\n",
       "      <td>1634.0</td>\n",
       "      <td>0.081425</td>\n",
       "      <td>False</td>\n",
       "      <td>1631.0</td>\n",
       "      <td>3102.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.354610</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.081425</td>\n",
       "      <td>False</td>\n",
       "      <td>1631.0</td>\n",
       "      <td>3102.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.354610</td>\n",
       "      <td>M-IL7R+</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>C8-Memory-1</td>\n",
       "      <td>BS1140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCAAACCAGGCAAGC-1-BS1140</th>\n",
       "      <td>977.0</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>False</td>\n",
       "      <td>976.0</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.642674</td>\n",
       "      <td>M-CD69+</td>\n",
       "      <td>M-CD69+</td>\n",
       "      <td>C6-EM-2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>False</td>\n",
       "      <td>976.0</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.642674</td>\n",
       "      <td>M-CD69+</td>\n",
       "      <td>C6-EM-2</td>\n",
       "      <td>C6-EM-2</td>\n",
       "      <td>BS1140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           GEX_n_genes  GEX_doublet_score  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140       2069.0           0.031680   \n",
       "TGAAACTGTTAGAGGG-1-BS1140        969.0           0.072093   \n",
       "TTGACATCAGTTTGTG-1-BS1140       1413.0           0.222222   \n",
       "TGCACTTGTTAGGTGC-1-BS1140       1634.0           0.081425   \n",
       "CCCAAACCAGGCAAGC-1-BS1140        977.0           0.021739   \n",
       "\n",
       "                           GEX_predicted_doublet  GEX_n_genes_by_counts  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140                  False                 2068.0   \n",
       "TGAAACTGTTAGAGGG-1-BS1140                  False                  967.0   \n",
       "TTGACATCAGTTTGTG-1-BS1140                  False                 1409.0   \n",
       "TGCACTTGTTAGGTGC-1-BS1140                  False                 1631.0   \n",
       "CCCAAACCAGGCAAGC-1-BS1140                  False                  976.0   \n",
       "\n",
       "                           GEX_total_counts  GEX_total_counts_mt  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140            4343.0                 11.0   \n",
       "TGAAACTGTTAGAGGG-1-BS1140            1523.0                 19.0   \n",
       "TTGACATCAGTTTGTG-1-BS1140            2666.0                 74.0   \n",
       "TGCACTTGTTAGGTGC-1-BS1140            3102.0                 11.0   \n",
       "CCCAAACCAGGCAAGC-1-BS1140            1556.0                 10.0   \n",
       "\n",
       "                           GEX_pct_counts_mt GEX_louvain_res_1.0 GEX_celltype  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140           0.253281             M-IL7R+      M-IL7R+   \n",
       "TGAAACTGTTAGAGGG-1-BS1140           1.247538                  Ex           Ex   \n",
       "TTGACATCAGTTTGTG-1-BS1140           2.775694             M-IL7R+      M-IL7R+   \n",
       "TGCACTTGTTAGGTGC-1-BS1140           0.354610             M-IL7R+      M-IL7R+   \n",
       "CCCAAACCAGGCAAGC-1-BS1140           0.642674             M-CD69+      M-CD69+   \n",
       "\n",
       "                          GEX_cluster_archr  ...  ACC_doublet_score  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140       C8-Memory-1  ...           0.031680   \n",
       "TGAAACTGTTAGAGGG-1-BS1140    C2-Exhausted-2  ...           0.072093   \n",
       "TTGACATCAGTTTGTG-1-BS1140       C8-Memory-1  ...           0.222222   \n",
       "TGCACTTGTTAGGTGC-1-BS1140       C8-Memory-1  ...           0.081425   \n",
       "CCCAAACCAGGCAAGC-1-BS1140           C6-EM-2  ...           0.021739   \n",
       "\n",
       "                           ACC_predicted_doublet ACC_n_genes_by_counts  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140                  False                2068.0   \n",
       "TGAAACTGTTAGAGGG-1-BS1140                  False                 967.0   \n",
       "TTGACATCAGTTTGTG-1-BS1140                  False                1409.0   \n",
       "TGCACTTGTTAGGTGC-1-BS1140                  False                1631.0   \n",
       "CCCAAACCAGGCAAGC-1-BS1140                  False                 976.0   \n",
       "\n",
       "                           ACC_total_counts  ACC_total_counts_mt  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140            4343.0                 11.0   \n",
       "TGAAACTGTTAGAGGG-1-BS1140            1523.0                 19.0   \n",
       "TTGACATCAGTTTGTG-1-BS1140            2666.0                 74.0   \n",
       "TGCACTTGTTAGGTGC-1-BS1140            3102.0                 11.0   \n",
       "CCCAAACCAGGCAAGC-1-BS1140            1556.0                 10.0   \n",
       "\n",
       "                           ACC_pct_counts_mt  ACC_louvain_res_1.0  \\\n",
       "ATTCCTCCATAATCAC-1-BS1140           0.253281              M-IL7R+   \n",
       "TGAAACTGTTAGAGGG-1-BS1140           1.247538                   Ex   \n",
       "TTGACATCAGTTTGTG-1-BS1140           2.775694              M-IL7R+   \n",
       "TGCACTTGTTAGGTGC-1-BS1140           0.354610              M-IL7R+   \n",
       "CCCAAACCAGGCAAGC-1-BS1140           0.642674              M-CD69+   \n",
       "\n",
       "                             ACC_celltype  ACC_cluster_archr  ACC_sample_id  \n",
       "ATTCCTCCATAATCAC-1-BS1140     C8-Memory-1        C8-Memory-1         BS1140  \n",
       "TGAAACTGTTAGAGGG-1-BS1140  C2-Exhausted-2     C2-Exhausted-2         BS1140  \n",
       "TTGACATCAGTTTGTG-1-BS1140     C8-Memory-1        C8-Memory-1         BS1140  \n",
       "TGCACTTGTTAGGTGC-1-BS1140     C8-Memory-1        C8-Memory-1         BS1140  \n",
       "CCCAAACCAGGCAAGC-1-BS1140         C6-EM-2            C6-EM-2         BS1140  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.metadata_cell.head() # cell metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001023</td>\n",
       "      <td>99.897716</td>\n",
       "      <td>3.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.006147</td>\n",
       "      <td>1.972477</td>\n",
       "      <td>1.508182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.5</th>\n",
       "      <td>ENSG00000241860</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>13</td>\n",
       "      <td>0.004773</td>\n",
       "      <td>99.556768</td>\n",
       "      <td>14.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.013088</td>\n",
       "      <td>1.195926</td>\n",
       "      <td>-0.803162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01409</th>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>198</td>\n",
       "      <td>False</td>\n",
       "      <td>197</td>\n",
       "      <td>0.077395</td>\n",
       "      <td>93.283328</td>\n",
       "      <td>227.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.190262</td>\n",
       "      <td>1.783858</td>\n",
       "      <td>0.946773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01128</th>\n",
       "      <td>ENSG00000228794</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>186</td>\n",
       "      <td>False</td>\n",
       "      <td>186</td>\n",
       "      <td>0.071258</td>\n",
       "      <td>93.658370</td>\n",
       "      <td>209.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.202416</td>\n",
       "      <td>1.536482</td>\n",
       "      <td>0.210478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>21</td>\n",
       "      <td>False</td>\n",
       "      <td>21</td>\n",
       "      <td>0.007160</td>\n",
       "      <td>99.284010</td>\n",
       "      <td>21.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.026549</td>\n",
       "      <td>1.773981</td>\n",
       "      <td>0.917376</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_ids    feature_types  genome  n_cells     mt  \\\n",
       "AL627309.1  ENSG00000238009  Gene Expression  GRCh38        3  False   \n",
       "AL627309.5  ENSG00000241860  Gene Expression  GRCh38       13  False   \n",
       "LINC01409   ENSG00000237491  Gene Expression  GRCh38      198  False   \n",
       "LINC01128   ENSG00000228794  Gene Expression  GRCh38      186  False   \n",
       "LINC00115   ENSG00000225880  Gene Expression  GRCh38       21  False   \n",
       "\n",
       "            n_cells_by_counts  mean_counts  pct_dropout_by_counts  \\\n",
       "AL627309.1                  3     0.001023              99.897716   \n",
       "AL627309.5                 13     0.004773              99.556768   \n",
       "LINC01409                 197     0.077395              93.283328   \n",
       "LINC01128                 186     0.071258              93.658370   \n",
       "LINC00115                  21     0.007160              99.284010   \n",
       "\n",
       "            total_counts  highly_variable     means  dispersions  \\\n",
       "AL627309.1           3.0            False  0.006147     1.972477   \n",
       "AL627309.5          14.0            False  0.013088     1.195926   \n",
       "LINC01409          227.0             True  0.190262     1.783858   \n",
       "LINC01128          209.0            False  0.202416     1.536482   \n",
       "LINC00115           21.0             True  0.026549     1.773981   \n",
       "\n",
       "            dispersions_norm  \n",
       "AL627309.1          1.508182  \n",
       "AL627309.5         -0.803162  \n",
       "LINC01409           0.946773  \n",
       "LINC01128           0.210478  \n",
       "LINC00115           0.917376  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.metadata_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Width</th>\n",
       "      <th>cisTopic_nr_frag</th>\n",
       "      <th>cisTopic_log_nr_frag</th>\n",
       "      <th>cisTopic_nr_acc</th>\n",
       "      <th>cisTopic_log_nr_acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GL000194.1:114742-115242</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>114742</td>\n",
       "      <td>115242</td>\n",
       "      <td>500</td>\n",
       "      <td>62</td>\n",
       "      <td>1.792392</td>\n",
       "      <td>62</td>\n",
       "      <td>1.792392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000194.1:101154-101654</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>101154</td>\n",
       "      <td>101654</td>\n",
       "      <td>500</td>\n",
       "      <td>244</td>\n",
       "      <td>2.387390</td>\n",
       "      <td>203</td>\n",
       "      <td>2.307496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000195.1:92476-92976</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>92476</td>\n",
       "      <td>92976</td>\n",
       "      <td>500</td>\n",
       "      <td>5</td>\n",
       "      <td>0.698970</td>\n",
       "      <td>5</td>\n",
       "      <td>0.698970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000195.1:131013-131513</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>131013</td>\n",
       "      <td>131513</td>\n",
       "      <td>500</td>\n",
       "      <td>19</td>\n",
       "      <td>1.278754</td>\n",
       "      <td>18</td>\n",
       "      <td>1.255273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000195.1:30584-31084</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>30584</td>\n",
       "      <td>31084</td>\n",
       "      <td>500</td>\n",
       "      <td>451</td>\n",
       "      <td>2.654177</td>\n",
       "      <td>400</td>\n",
       "      <td>2.602060</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Chromosome   Start     End  Width  cisTopic_nr_frag  \\\n",
       "GL000194.1:114742-115242  GL000194.1  114742  115242    500                62   \n",
       "GL000194.1:101154-101654  GL000194.1  101154  101654    500               244   \n",
       "GL000195.1:92476-92976    GL000195.1   92476   92976    500                 5   \n",
       "GL000195.1:131013-131513  GL000195.1  131013  131513    500                19   \n",
       "GL000195.1:30584-31084    GL000195.1   30584   31084    500               451   \n",
       "\n",
       "                          cisTopic_log_nr_frag  cisTopic_nr_acc  \\\n",
       "GL000194.1:114742-115242              1.792392               62   \n",
       "GL000194.1:101154-101654              2.387390              203   \n",
       "GL000195.1:92476-92976                0.698970                5   \n",
       "GL000195.1:131013-131513              1.278754               18   \n",
       "GL000195.1:30584-31084                2.654177              400   \n",
       "\n",
       "                          cisTopic_log_nr_acc  \n",
       "GL000194.1:114742-115242             1.792392  \n",
       "GL000194.1:101154-101654             2.307496  \n",
       "GL000195.1:92476-92976               0.698970  \n",
       "GL000195.1:131013-131513             1.255273  \n",
       "GL000195.1:30584-31084               2.602060  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.metadata_regions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Motif enrichment data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['CTX_topics_otsu_All', 'CTX_topics_otsu_No_promoters', 'DEM_topics_otsu_All', 'DEM_topics_otsu_No_promoters', 'CTX_topics_top_3_All', 'CTX_topics_top_3_No_promoters', 'DEM_topics_top_3_All', 'DEM_topics_top_3_No_promoters', 'CTX_DARs_All', 'CTX_DARs_No_promoters', 'DEM_DARs_All', 'DEM_DARs_No_promoters'])"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.menr.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dimensionality reduction data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['GEX_X_pca', 'GEX_X_umap', 'eRegulons_UMAP', 'eRegulons_tSNE'])"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.dr_cell.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "some unstructured info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Cistromes', 'search_space', 'region_to_gene', 'TF2G_adj', 'eRegulons', 'eRegulon_metadata', 'eRegulon_signatures', 'eRegulon_AUC', 'Pseudobulk', 'TF_cistrome_correlation', 'eRegulon_AUC_thresholds', 'RSS', 'DEGs', 'DARs'])"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.uns.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The eRegulons entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[eRegulon for TF ARID5B in context frozenset({'positive tf2g', 'Top 15 region-to-gene links per gene', 'positive r2g', 'Top 10 region-to-gene links per gene', 'Cistromes_Unfiltered', 'BASC binarized'}).\n",
       " \tThis eRegulon has 19 target regions and 15 target genes.,\n",
       " eRegulon for TF ATF2 in context frozenset({'positive tf2g', 'Top 15 region-to-gene links per gene', 'positive r2g', 'Top 10 region-to-gene links per gene', 'Cistromes_Unfiltered', 'BASC binarized', '0.85 quantile'}).\n",
       " \tThis eRegulon has 18 target regions and 13 target genes.,\n",
       " eRegulon for TF ATF3 in context frozenset({'positive tf2g', 'Top 15 region-to-gene links per gene', 'positive r2g', 'Top 10 region-to-gene links per gene', 'Cistromes_Unfiltered', 'BASC binarized', 'Top 5 region-to-gene links per gene', '0.85 quantile', '0.9 quantile', '0.95 quantile'}).\n",
       " \tThis eRegulon has 118 target regions and 51 target genes.,\n",
       " eRegulon for TF ATF6 in context frozenset({'positive tf2g', 'Cistromes_Unfiltered', 'BASC binarized', 'positive r2g'}).\n",
       " \tThis eRegulon has 14 target regions and 11 target genes.,\n",
       " eRegulon for TF BACH1 in context frozenset({'positive tf2g', 'Top 15 region-to-gene links per gene', 'positive r2g', 'Top 10 region-to-gene links per gene', 'Cistromes_Unfiltered', 'BASC binarized', 'Top 5 region-to-gene links per gene', '0.85 quantile', '0.9 quantile', '0.95 quantile'}).\n",
       " \tThis eRegulon has 256 target regions and 95 target genes.]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.uns['eRegulons'][0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cistrome_name: ARID5B_(517r)\n",
      "context: frozenset({'positive tf2g', 'Top 15 region-to-gene links per gene', 'positive r2g', 'Top 10 region-to-gene links per gene', 'Cistromes_Unfiltered', 'BASC binarized'})\n",
      "gsea_adj_pval: None\n",
      "gsea_enrichment_score: None\n",
      "gsea_pval: None\n",
      "in_leading_edge: None\n",
      "is_extended: False\n",
      "n_target_genes: 15\n",
      "n_target_regions: 19\n",
      "regions2genes: [r2g(region='chr5:157344232-157344732', target='ITK', importance=0.01694017203078059, rho=0.19743390222890106, importance_x_rho=0.003344564268465899, importance_x_abs_rho=0.003344564268465899), r2g(region='chr6:16473114-16473614', target='ATXN1', importance=0.036620395425374805, rho=0.33063230171542995, importance_x_rho=0.012107885629220874, importance_x_abs_rho=0.012107885629220874), r2g(region='chr2:203861321-203861821', target='ICOS', importance=0.030269561376561387, rho=0.3582964469743678, importance_x_rho=0.010845476292694499, importance_x_abs_rho=0.010845476292694499), r2g(region='chr2:203914728-203915228', target='ICOS', importance=0.019689863504460038, rho=0.3011347745975594, importance_x_rho=0.005929302608272284, importance_x_abs_rho=0.005929302608272284), r2g(region='chr8:19687777-19688277', target='CSGALNACT1', importance=0.05486208575662483, rho=0.5906063941361934, importance_x_rho=0.032401898643510806, importance_x_abs_rho=0.032401898643510806)]\n",
      "subset_leading_edge: <bound method eRegulon.subset_leading_edge of eRegulon for TF ARID5B in context frozenset({'positive tf2g', 'Top 15 region-to-gene links per gene', 'positive r2g', 'Top 10 region-to-gene links per gene', 'Cistromes_Unfiltered', 'BASC binarized'}).\n",
      "\tThis eRegulon has 19 target regions and 15 target genes.>\n",
      "target_genes: ['LYST', 'HNRNPLL', 'ZNF831', 'NAB1', 'TTN']\n",
      "target_regions: ['chr2:38685151-38685651', 'chr6:16419719-16420219', 'chr2:178404262-178404762', 'chr2:203861321-203861821', 'chr2:190692669-190693169']\n",
      "transcription_factor: ARID5B\n"
     ]
    }
   ],
   "source": [
    "for attr in dir(scplus_obj.uns['eRegulons'][0]):\n",
    "    if not attr.startswith('_'):\n",
    "        print(f\"{attr}: {getattr(scplus_obj.uns['eRegulons'][0], attr) if not type(getattr(scplus_obj.uns['eRegulons'][0], attr)) == list else getattr(scplus_obj.uns['eRegulons'][0], attr)[0:5]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region_signature_name</th>\n",
       "      <th>Gene_signature_name</th>\n",
       "      <th>TF</th>\n",
       "      <th>is_extended</th>\n",
       "      <th>Region</th>\n",
       "      <th>Gene</th>\n",
       "      <th>R2G_importance</th>\n",
       "      <th>R2G_rho</th>\n",
       "      <th>R2G_importance_x_rho</th>\n",
       "      <th>R2G_importance_x_abs_rho</th>\n",
       "      <th>TF2G_importance</th>\n",
       "      <th>TF2G_regulation</th>\n",
       "      <th>TF2G_rho</th>\n",
       "      <th>TF2G_importance_x_abs_rho</th>\n",
       "      <th>TF2G_importance_x_rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARID5B_+_+_(19r)</td>\n",
       "      <td>ARID5B_+_+_(15g)</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>False</td>\n",
       "      <td>chr5:157344232-157344732</td>\n",
       "      <td>ITK</td>\n",
       "      <td>0.016940</td>\n",
       "      <td>0.197434</td>\n",
       "      <td>0.003345</td>\n",
       "      <td>0.003345</td>\n",
       "      <td>9.205713</td>\n",
       "      <td>1</td>\n",
       "      <td>0.115597</td>\n",
       "      <td>1.064154</td>\n",
       "      <td>1.064154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID5B_+_+_(19r)</td>\n",
       "      <td>ARID5B_+_+_(15g)</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>False</td>\n",
       "      <td>chr6:16473114-16473614</td>\n",
       "      <td>ATXN1</td>\n",
       "      <td>0.036620</td>\n",
       "      <td>0.330632</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>0.012108</td>\n",
       "      <td>1.737764</td>\n",
       "      <td>1</td>\n",
       "      <td>0.111201</td>\n",
       "      <td>0.193241</td>\n",
       "      <td>0.193241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARID5B_+_+_(19r)</td>\n",
       "      <td>ARID5B_+_+_(15g)</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>False</td>\n",
       "      <td>chr6:16419719-16420219</td>\n",
       "      <td>ATXN1</td>\n",
       "      <td>0.015978</td>\n",
       "      <td>0.325449</td>\n",
       "      <td>0.005200</td>\n",
       "      <td>0.005200</td>\n",
       "      <td>1.737764</td>\n",
       "      <td>1</td>\n",
       "      <td>0.111201</td>\n",
       "      <td>0.193241</td>\n",
       "      <td>0.193241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARID5B_+_+_(19r)</td>\n",
       "      <td>ARID5B_+_+_(15g)</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>False</td>\n",
       "      <td>chr6:16699292-16699792</td>\n",
       "      <td>ATXN1</td>\n",
       "      <td>0.004254</td>\n",
       "      <td>0.159969</td>\n",
       "      <td>0.000680</td>\n",
       "      <td>0.000680</td>\n",
       "      <td>1.737764</td>\n",
       "      <td>1</td>\n",
       "      <td>0.111201</td>\n",
       "      <td>0.193241</td>\n",
       "      <td>0.193241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARID5B_+_+_(19r)</td>\n",
       "      <td>ARID5B_+_+_(15g)</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>False</td>\n",
       "      <td>chr2:203861321-203861821</td>\n",
       "      <td>ICOS</td>\n",
       "      <td>0.030270</td>\n",
       "      <td>0.358296</td>\n",
       "      <td>0.010845</td>\n",
       "      <td>0.010845</td>\n",
       "      <td>4.094361</td>\n",
       "      <td>1</td>\n",
       "      <td>0.099097</td>\n",
       "      <td>0.405741</td>\n",
       "      <td>0.405741</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Region_signature_name Gene_signature_name      TF is_extended  \\\n",
       "0      ARID5B_+_+_(19r)    ARID5B_+_+_(15g)  ARID5B       False   \n",
       "1      ARID5B_+_+_(19r)    ARID5B_+_+_(15g)  ARID5B       False   \n",
       "2      ARID5B_+_+_(19r)    ARID5B_+_+_(15g)  ARID5B       False   \n",
       "3      ARID5B_+_+_(19r)    ARID5B_+_+_(15g)  ARID5B       False   \n",
       "4      ARID5B_+_+_(19r)    ARID5B_+_+_(15g)  ARID5B       False   \n",
       "\n",
       "                     Region   Gene  R2G_importance   R2G_rho  \\\n",
       "0  chr5:157344232-157344732    ITK        0.016940  0.197434   \n",
       "1    chr6:16473114-16473614  ATXN1        0.036620  0.330632   \n",
       "2    chr6:16419719-16420219  ATXN1        0.015978  0.325449   \n",
       "3    chr6:16699292-16699792  ATXN1        0.004254  0.159969   \n",
       "4  chr2:203861321-203861821   ICOS        0.030270  0.358296   \n",
       "\n",
       "   R2G_importance_x_rho  R2G_importance_x_abs_rho  TF2G_importance  \\\n",
       "0              0.003345                  0.003345         9.205713   \n",
       "1              0.012108                  0.012108         1.737764   \n",
       "2              0.005200                  0.005200         1.737764   \n",
       "3              0.000680                  0.000680         1.737764   \n",
       "4              0.010845                  0.010845         4.094361   \n",
       "\n",
       "   TF2G_regulation  TF2G_rho  TF2G_importance_x_abs_rho  TF2G_importance_x_rho  \n",
       "0                1  0.115597                   1.064154               1.064154  \n",
       "1                1  0.111201                   0.193241               0.193241  \n",
       "2                1  0.111201                   0.193241               0.193241  \n",
       "3                1  0.111201                   0.193241               0.193241  \n",
       "4                1  0.099097                   0.405741               0.405741  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scplus_obj.uns['eRegulon_metadata'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export cell metadata\n",
    "scplus_obj.uns['eRegulon_metadata'].to_csv('/Users/jinhuixin/Master/thesis/GRN/BS1140/eRegulon_metadata.csv')"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
